src/error_diffusion.jl

```diff
@@ -44,41 +55,37 @@ function error_diffusion(
     return dither
 end
 
-"""
-Get indices from OffsetMatrix
-"""
-function indices(om::OffsetMatrix)
-    rows, cols = size(om)
-    row_range = (1:rows) .+ om.offsets[1]
-    col_range = (1:cols) .+ om.offsets[2]
-    return row_range, col_range
+function dither(
+    img::AbstractMatrix{T}, alg::ErrorDiffusion, cs::ColorScheme; kwargs...
+)::AbstractMatrix{T} where {T<:Color}
+    return dither(img, alg, cs.colors; kwargs...)
 end
 
-simple_error_diffusion(img; kwargs...) = error_diffusion(img, SIMPLE_STENCIL; kwargs...)
-const SIMPLE_STENCIL = OffsetArray([0 1; 1 0]//2, 0:1, 0:1)
-
-floyd_steinberg_diffusion(img; kwargs...) = error_diffusion(img, FS_STENCIL; kwargs...)
-const FS_STENCIL = OffsetArray([0 0 7; 3 5 1]//16, 0:1, -1:1)
-
-jarvis_judice_diffusion(img; kwargs...) = error_diffusion(img, JJ_STENCIL; kwargs...)
-const JJ_STENCIL = OffsetArray([0 0 0 7 5; 3 5 7 5 3; 1 3 5 3 1]//48, 0:2, -2:2)
-
-stucki_diffusion(img; kwargs...) = error_diffusion(img, STUCKI_STENCIL; kwargs...)
-const STUCKI_STENCIL = OffsetArray([0 0 0 8 4; 2 4 8 4 2; 1 2 4 2 1]//42, 0:2, -2:2)
-
-burkes_diffusion(img; kwargs...) = error_diffusion(img, BURKES_STENCIL; kwargs...)
-const BURKES_STENCIL = OffsetArray([0 0 0 8 4; 2 4 8 4 2]//32, 0:1, -2:2)
-
-sierra_diffusion(img; kwargs...) = error_diffusion(img, SIERRA_STENCIL; kwargs...)
-const SIERRA_STENCIL = OffsetArray([0 0 0 5 3; 2 4 5 4 2; 0 2 3 2 0]//32, 0:2, -2:2)
-
-function two_row_sierra_diffusion(img; kwargs...)
-    return error_diffusion(img, TWO_ROW_SIERRA_STENCIL; kwargs...)
+function dither(
+    img::AbstractMatrix{<:AbstractGray},
+    alg::ErrorDiffusion;
+    steps=2,
+    metric=BinaryDitherMetric(),
+    kwargs...,
+)
+    # Construct grayscale color scheme of length `steps`
+    cs = [Gray(i) for i in range(0.0, 1.0; length=steps)]
+    d = dither(img, alg, cs; metric, kwargs...)
+    if steps == 2
+        # match return type of other binary dithering algorithms
+        return Gray{Bool}.(d)
+    end
+    return d
```

Comment:
This is not type-stable and would hurt the performance of its consumer:

```suggestion
    return reinterpret(Gray{Bool}, d)
```

```julia
function f(; step=2)
    x = rand(Bool, 4, 4)
    if step == 2
        return Gray{Bool}.(x)
    end
    return x
end

# you can check the return type with `@code_warntype`
@code_warntype f()
@code_warntype f(step=3)
```

Reply:
Somehow `reinterpret` added a bunch of rows to the output.
@johnnychen94 I think this might warrant an issue! 
```julia
julia> using ImageCore

julia> A = Gray{Float32}.([1 1; 0 0])
2×2 Array{Gray{Float32},2} with eltype Gray{Float32}:
 Gray{Float32}(1.0)  Gray{Float32}(1.0)
 Gray{Float32}(0.0)  Gray{Float32}(0.0)

julia> reinterpret(Gray{Bool}, A)
8×2 reinterpret(Gray{Bool}, ::Array{Gray{Float32},2}):
 Gray{Bool}(false)  Gray{Bool}(false)
 Gray{Bool}(false)  Gray{Bool}(false)
 Gray{Bool}(true)   Gray{Bool}(true)
 Gray{Bool}(true)   Gray{Bool}(true)
 Gray{Bool}(false)  Gray{Bool}(false)
 Gray{Bool}(false)  Gray{Bool}(false)
 Gray{Bool}(false)  Gray{Bool}(false)
 Gray{Bool}(false)  Gray{Bool}(false)
 ```
 I have reverted `reinterpret` for now.

Reply:
Ahh, I didn't realize that there can be float types here. `reinterpret` works like C/C++ pointer interpolation, so storage layout matters. Maybe:

```julia
return convert(Marix{Gray{Bool}}, d)
```

When `d` is already a `Matrix{Gray{Bool}}` type, it does nothing; otherwise it calls some conversion function to create it (via copy constructor or others).